### Fine Tuning of bi-encoder and cross encoder

В этом ноутбуке мы делаем тюнинг bi-энкодера и cross-энкодера

#### 1. Fine Tuning of bi-encoder

In [11]:
import numpy as np
import pandas as pd

# импотируем свой модуль, не внося постоянные изменения в path
import sys
sys.path.append("C:\\Users\\satyr\\Documents\\edu\\nlp2\\hw1\\")
from my_tokenize_vectorize import Tokenizer, SentBERT_vectorizer

tk = Tokenizer('dummy', lower = False)
v = SentBERT_vectorizer('all-mpnet-base-v2')
print(tk.tokenize_corpus(corpus=['test! test... Test', np.nan]))
print(v.vectorize_tokens(tk.tokenize_corpus(corpus=['test! test... Test', np.nan])))
# видим, что приведение к нижнему регистру и токенизация работают и вакторизация работают

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\satyr\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


['test! test... Test', nan]
[[-0.02506025  0.03046899 -0.0486784  ...  0.01635451 -0.03483407
  -0.01601352]
 [-0.02318716  0.05149755 -0.00239227 ...  0.02563848 -0.07033019
  -0.00896621]]


In [12]:
aa = pd.read_pickle('data/eng_script_vectorized_v2.pkl')

In [13]:
aa.columns

Index(['series_num', 'ep_num', 'ep_name', 'phrase_rank', 'person', 'text',
       'person_orig', 'person_context', 'scene_context', 'text_5_prev',
       'person_5_prev', 'person_2_prev', 'text_1_shift', 'text_2_shift',
       'text_3_shift', 'text_4_shift', 'text_5_shift', 'glove_v_text',
       'glove_v_text_1_shift', 'glove_v_text_2_shift', 'glove_v_text_3_shift',
       'glove_v_text_4_shift', 'glove_v_text_5_shift', 'sbert_v_text_1_shift',
       'sbert_v_text', 'sbert_v_text_2_shift', 'sbert_v_text_3_shift',
       'sbert_v_text_4_shift', 'sbert_v_text_5_shift'],
      dtype='object')

In [14]:
anchor = aa[aa.person == 'Leonard'].text_1_shift.apply(lambda x: str(x)).to_list()
positive = aa[aa.person == 'Leonard'].text.apply(lambda x: str(x)).to_list()
assert len(anchor) == len(positive)
_ = len(anchor)
# Sheldon is counterpart to Leonard
negative = aa[aa.person == 'Sheldon'].text.apply(lambda x: str(x)).to_list()
n_temp = len(anchor)
negative = (negative + n_temp*[''])[:n_temp]
assert len(anchor) == len(negative)

In [15]:
from sentence_transformers.readers import InputExample
from torch.utils.data import DataLoader

In [16]:
train_examples = []
for i in range(min(1000, n_temp)):
    train_examples.append(InputExample(texts=[anchor[i], positive[i], negative[i]]))
print(train_examples[:5])
train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=8)


[<sentence_transformers.readers.InputExample.InputExample object at 0x000001CB847612D0>, <sentence_transformers.readers.InputExample.InputExample object at 0x000001CB84781890>, <sentence_transformers.readers.InputExample.InputExample object at 0x000001CB847815D0>, <sentence_transformers.readers.InputExample.InputExample object at 0x000001CBA3B2A1D0>, <sentence_transformers.readers.InputExample.InputExample object at 0x000001CBA7829550>]


In [12]:
from sentence_transformers import losses, util
from sentence_transformers import SentenceTransformer
v = SentenceTransformer('all-mpnet-base-v2')
train_loss = losses.TripletLoss(model=v)
v.fit(train_objectives=[(train_dataloader, train_loss)], epochs=1) 


In [8]:
v.save("model/triple-1000-fit-all-mpnet-base-v2")

In [9]:
v = SentenceTransformer('model/triple-fit-all-mpnet-base-v2')
v1 = v.encode(['Hello how are you? Want to go to the uni?'])[0]

In [10]:
v = SentenceTransformer('all-mpnet-base-v2')
v2 = v.encode(['Hello how are you? Want to go to the bar?'])[0]

In [11]:
from sentence_transformers import util
util.cos_sim(v1, v2)

tensor([[0.4937]])

In [15]:
bi_orig_model = SentenceTransformer('all-mpnet-base-v2')
pos_cos_sim, neg_cos_sim, n_max = [], [], 200
for i in range(n_max):
    pos_cos_sim.append(util.cos_sim(bi_orig_model.encode(anchor[i]), bi_orig_model.encode(positive[i])))
    neg_cos_sim.append(util.cos_sim(bi_orig_model.encode(anchor[i]), bi_orig_model.encode(negative[i])))
print('BEFOR pos_cos_sim', sum(pos_cos_sim)/len(pos_cos_sim), 'neg_cos_sim', sum(neg_cos_sim)/len(neg_cos_sim), sep='\t')

AFTER pos_cos_sim	tensor([[0.2241]])	neg_cos_sim	tensor([[0.1805]])


In [14]:
bi_ft_model = SentenceTransformer('model/triple-1000-fit-all-mpnet-base-v2')
pos_cos_sim, neg_cos_sim, n_max = [], [], 200
for i in range(n_max):
    pos_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(positive[i])))
    neg_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(negative[i])))
print('AFTER pos_cos_sim', sum(pos_cos_sim)/len(pos_cos_sim), 'neg_cos_sim', sum(neg_cos_sim)/len(neg_cos_sim), sep='\t')

AFTER pos_cos_sim	tensor([[0.2251]])	neg_cos_sim	tensor([[0.1795]])


Сравнили средине косинусные близости позитивных и негативных примеров до и после:
 * Before pos_cos_sim	tensor([[0.2241]])	neg_cos_sim	tensor([[0.1805]])
 * AFTER pos_cos_sim	tensor([[0.2251]])	neg_cos_sim	tensor([[0.1795]])

видим, что разница есть, но не большая, поэтому сделаем еще 5 эпох обучения:

 * AFTER pos_cos_sim	tensor([[0.2675]])	neg_cos_sim	tensor([[0.1744]])

разница растет слабо, добавим еще 6 эпох:
 * AFTER pos_cos_sim	tensor([[0.3534]])	neg_cos_sim	tensor([[0.1976]])

и еще 8 эпох:
 * AFTER pos_cos_sim	tensor([[0.5012]])	neg_cos_sim	tensor([[0.2178]])

На этом остановимся.

In [9]:
pwd

'C:\\Users\\satyr\\Documents\\edu\\nlp2\\hw1'

In [17]:
from sentence_transformers import losses, util
from sentence_transformers import SentenceTransformer

bi_ft_model = SentenceTransformer("model\\triple-1000-fit-all-mpnet-base-v2")
train_loss = losses.TripletLoss(model=bi_ft_model)
bi_ft_model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=5) 
bi_ft_model.save("model\\triple-6e-1000-fit-all-mpnet-base-v2")

Epoch:   0%|          | 0/5 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

In [18]:
bi_ft_model = SentenceTransformer('model\\triple-6e-1000-fit-all-mpnet-base-v2')
pos_cos_sim, neg_cos_sim, n_max = [], [], 200
for i in range(n_max):
    pos_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(positive[i])))
    neg_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(negative[i])))
print('AFTER pos_cos_sim', sum(pos_cos_sim)/len(pos_cos_sim), 'neg_cos_sim', sum(neg_cos_sim)/len(neg_cos_sim), sep='\t')

AFTER pos_cos_sim	tensor([[0.2675]])	neg_cos_sim	tensor([[0.1744]])


In [19]:
bi_ft_model = SentenceTransformer("model\\triple-6e-1000-fit-all-mpnet-base-v2")
train_loss = losses.TripletLoss(model=bi_ft_model)
bi_ft_model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=6) 
bi_ft_model.save("model\\triple-12e-1000-fit-all-mpnet-base-v2")

Epoch:   0%|          | 0/6 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

In [20]:
bi_ft_model = SentenceTransformer('model\\triple-12e-1000-fit-all-mpnet-base-v2')
pos_cos_sim, neg_cos_sim, n_max = [], [], 200
for i in range(n_max):
    pos_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(positive[i])))
    neg_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(negative[i])))
print('AFTER pos_cos_sim', sum(pos_cos_sim)/len(pos_cos_sim), 'neg_cos_sim', sum(neg_cos_sim)/len(neg_cos_sim), sep='\t')

AFTER pos_cos_sim	tensor([[0.3534]])	neg_cos_sim	tensor([[0.1976]])


In [21]:
bi_ft_model = SentenceTransformer("model\\triple-12e-1000-fit-all-mpnet-base-v2")
train_loss = losses.TripletLoss(model=bi_ft_model)
bi_ft_model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=8) 
bi_ft_model.save("model\\triple-20e-1000-fit-all-mpnet-base-v2")

Epoch:   0%|          | 0/8 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

Iteration:   0%|          | 0/125 [00:00<?, ?it/s]

In [22]:
bi_ft_model = SentenceTransformer('model\\triple-20e-1000-fit-all-mpnet-base-v2')
pos_cos_sim, neg_cos_sim, n_max = [], [], 200
for i in range(n_max):
    pos_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(positive[i])))
    neg_cos_sim.append(util.cos_sim(bi_ft_model.encode(anchor[i]), bi_ft_model.encode(negative[i])))
print('AFTER pos_cos_sim', sum(pos_cos_sim)/len(pos_cos_sim), 'neg_cos_sim', sum(neg_cos_sim)/len(neg_cos_sim), sep='\t')

AFTER pos_cos_sim	tensor([[0.5012]])	neg_cos_sim	tensor([[0.2178]])


In [1]:
!pip install huggingface_hub

In [5]:
# сохраним модель и на хаб тож:
from huggingface_hub import notebook_login
# token: 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
notebook_login()
from sentence_transformers import SentenceTransformer
bi_ft_model = SentenceTransformer("model\\triple-20e-1000-fit-all-mpnet-base-v2")

pytorch_model.bin:   0%|          | 0.00/438M [00:00<?, ?B/s]

Upload 1 LFS files:   0%|          | 0/1 [00:00<?, ?it/s]

'https://huggingface.co/satyroffrost/triple-20e-1000-fit-all-mpnet-base-v2/commit/3e32856dc19cb2c6762b71ca0000df547e9612c4'

In [57]:
# read local model
bi_ft_model = SentenceTransformer("model\\triple-20e-1000-fit-all-mpnet-base-v2")
print('local', bi_ft_model.encode([("Push model to HuggingFace")])[0][:5])
# push local model to hub (tokenizer would not be uploaded to HF)
bi_ft_model.save_to_hub("satyroffrost/triple-20e-1000-fit-all-mpnet-base-v2", exist_ok=True)
# check if HF model works:
bi_ft_model = SentenceTransformer("satyroffrost/triple-20e-1000-fit-all-mpnet-base-v2")
print('hf model', bi_ft_model.encode([("Push model to HuggingFace")])[0][:5])

local [-0.01058503 -0.06358196  0.02812308  0.00658665  0.01420205]


hf model [-0.01058503 -0.06358196  0.02812308  0.00658665  0.01420205]


In [ ]:
bi_ft_model.sa

### 2. Fine Tuning of cross-encoder

Для треникровки этой модели мы испольузем специальную функцию потерь (на самом деле evaluator, а не loss) CERerankingEvaluator, встроенную в моделу sentence transformers.

Для дообучения кросс-энкоддеров не реализована функция потерь через косинусную близость, см ответ разработчика на мой вопрос на эту тему на гитхабе
https://github.com/UKPLab/sentence-transformers/issues/2483

In [31]:
import numpy as np
import pandas as pd

from sentence_transformers import InputExample, CrossEncoder, losses
from sentence_transformers.cross_encoder import CrossEncoder
from sentence_transformers.cross_encoder.evaluation import CERerankingEvaluator
from torch.utils.data import DataLoader

In [13]:
aa = pd.read_pickle('data/eng_script_vectorized_v2.pkl')

In [14]:
anchor = aa[aa.person == 'Leonard'].text_1_shift.apply(lambda x: str(x)).to_list()
positive = aa[aa.person == 'Leonard'].text.apply(lambda x: str(x)).to_list()
assert len(anchor) == len(positive)
_ = len(anchor)
# Sheldon is counterpart to Leonard
negative = aa[aa.person == 'Sheldon'].text.apply(lambda x: str(x)).to_list()
n_temp = len(anchor)
negative = (negative + n_temp*[''])[:n_temp]
assert len(anchor) == len(negative)

In [36]:
samples = [] # из документации: Must be a list and each element is of the form: {'query': '', 'positive': [], 'negative': []}
for i in range(len(anchor)):
    sample = {'query': anchor[i], 'positive': [positive[i]], 'negative': [negative[i]]}
    samples.append(sample)
samples[:5]

[{'query': 'nan',
  'positive': ['See, the liquid metal Terminators were created in the future by Skynet, and Skynet was developed by Miles Dyson, but that future no longer exists, due to Dyson’s death in Terminator 2.'],
  'negative': ['Okay, then riddle me this. Assuming all the good Terminators were originally evil Terminators created by Skynet but then reprogrammed by the future John Connor, why would Skynet, an artificial computer intelligence, bother to create a petite hot 17 year-old killer robot?']},
 {'query': 'Okay, then riddle me this. Assuming all the good Terminators were originally evil Terminators created by Skynet but then reprogrammed by the future John Connor, why would Skynet, an artificial computer intelligence, bother to create a petite hot 17 year-old killer robot?',
  'positive': ['Skynet is kinky? I don’t know. '],
  'negative': ['Artificial intelligences do not have teen fetishes. ']},
 {'query': 'Artificial intelligences do not have teen fetishes. ',
  'positi

In [43]:
train_examples = []
# add positive examples
for i in range(min(500, n_temp)):
    train_examples.append(InputExample(texts=[anchor[i], positive[i]], label=1.0))
# add negative examples
for i in range(min(5000, n_temp)):
    train_examples.append(InputExample(texts=[anchor[i], negative[i]], label=0.0))
print(len(train_examples))
train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=8)

4000


In [77]:
model = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2', num_labels=1, max_length=512)

In [47]:
# curr_loss = losses.CosineSimilarityLoss(model=model)
eval = CERerankingEvaluator(samples)

In [78]:
# model.fit(train_dataloader=train_dataloader, epochs=1, loss_fct=curr_loss)
model.fit(train_dataloader=train_dataloader, epochs=3, evaluator=eval)

Epoch:   0%|          | 0/3 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

In [79]:
model.save("model/crerankingeval-3e-4000-ms-marco-MiniLM-L-6-v2")

#### Оценим, дообучилась ли наша модель

для этого посчитаем средние скоры положительных и отрицательных ответов до и после обучения

Результаты одной эпохи (4000 обоазцов)
 * BEFORE positive avg score	-7.840192528441548	nagative avg score	-8.286745544821024
 * AFTER  positive avg score	-0.8207377856783569	nagative avg score	-2.355241816714406

еще 2 эпохи
 * 0.830748250540346	nagative avg score	-0.033121718301554214

еще эпохи:
 * AFTER  positive avg score	0.9061768487049267	nagative avg score	-1.5693068397347816

Результат после еще200 эпох
 * AFTER  positive avg score	10.22114202260971	nagative avg score	-10.5234565496444

В сумме обучение заняло 30 эпох и правильнее было бы реализовать коллбэк, который следи за изменением скоров и просто прервать обучение, когда разница будет существеннйой7

In [80]:
model = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2', num_labels=1, max_length=512)
print("Model prediction BEFORE fine tuning")
# for q in samples[:5]:
#     pos_pred = model.predict([[q['query'], q['positive'][0] ]])[0]
#     neg_pred = model.predict([[q['query'], q['negative'][0] ]])[0]
#     print(pos_pred, neg_pred, q['query'][:20], q['positive'][0][:20], q['negative'][0][:20], sep='\t')

pos_score_list, neg_score_list, n_samp = [], [], 200
for q in samples[:n_samp]:
    pos_score_list.append(model.predict([[q['query'], q['positive'][0]]])[0])
    neg_score_list.append(model.predict([[q['query'], q['negative'][0]]])[0])

print('BEFORE positive avg score', sum(pos_score_list)/len(pos_score_list), 'nagative avg score', sum(neg_score_list)/len(neg_score_list), sep='\t')

Model prediction BEFORE fine tuning
BEFORE positive avg score	-7.840192528441548	nagative avg score	-8.286745544821024


In [81]:
ft_model = CrossEncoder("model/crerankingeval-3e-4000-ms-marco-MiniLM-L-6-v2")
print("Model prediction AFTER fine tuning")
# for q in samples[:5]:
#     pos_pred = ft_model.predict([[q['query'], q['positive'][0] ]])[0]
#     neg_pred = ft_model.predict([[q['query'], q['negative'][0] ]])[0]
#     print(pos_pred, neg_pred, q['query'][:20], q['positive'][0][:20], q['negative'][0][:20], sep='\t')
    
pos_score_list, neg_score_list, n_samp = [], [], 200
for q in samples[:n_samp]:
    pos_score_list.append(ft_model.predict([[q['query'], q['positive'][0]]])[0])
    neg_score_list.append(ft_model.predict([[q['query'], q['negative'][0] ]])[0])

print('AFTER  positive avg score', sum(pos_score_list)/len(pos_score_list), 'nagative avg score', sum(neg_score_list)/len(neg_score_list), sep='\t')

Model prediction AFTER fine tuning
AFTER  positive avg score	0.6320126322493889	nagative avg score	-0.06215588595252484


продолжим дообучать модель, еще две эпохи:

In [82]:
# model.fit(train_dataloader=train_dataloader, epochs=1, loss_fct=curr_loss)
ft_model.fit(train_dataloader=train_dataloader, epochs=2, evaluator=eval)

Epoch:   0%|          | 0/2 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

In [83]:
pos_score_list, neg_score_list, n_samp = [], [], 200
for q in samples[:n_samp]:
    pos_score_list.append(ft_model.predict([[q['query'], q['positive'][0]]])[0])
    neg_score_list.append(ft_model.predict([[q['query'], q['negative'][0] ]])[0])

print('AFTER  positive avg score', sum(pos_score_list)/len(pos_score_list), 'nagative avg score', sum(neg_score_list)/len(neg_score_list), sep='\t')

AFTER  positive avg score	0.830748250540346	nagative avg score	-0.033121718301554214


In [ ]:
продолжим дообучать модель, еще пять эпох:

In [84]:
# model.fit(train_dataloader=train_dataloader, epochs=1, loss_fct=curr_loss)
ft_model.fit(train_dataloader=train_dataloader, epochs=5, evaluator=eval)

Epoch:   0%|          | 0/5 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

In [86]:
ft_model.save("model/crerankingeval-10e-4000-ms-marco-MiniLM-L-6-v2")

In [87]:
pos_score_list, neg_score_list, n_samp = [], [], 200
for q in samples[:n_samp]:
    pos_score_list.append(ft_model.predict([[q['query'], q['positive'][0]]])[0])
    neg_score_list.append(ft_model.predict([[q['query'], q['negative'][0] ]])[0])

print('AFTER  positive avg score', sum(pos_score_list)/len(pos_score_list), 'nagative avg score', sum(neg_score_list)/len(neg_score_list), sep='\t')

AFTER  positive avg score	0.9061768487049267	nagative avg score	-1.5693068397347816


видим, что с ростом числа эпох разделение негативных и поситивных примеров по среднему скору имеет место быть,
поставим на ночь обучение еще на 20 эпох:

In [88]:
ft_model.fit(train_dataloader=train_dataloader, epochs=20, evaluator=eval)
ft_model.save("model/crerankingeval-30e-4000-ms-marco-MiniLM-L-6-v2")

Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

Iteration:   0%|          | 0/500 [00:00<?, ?it/s]

In [89]:
pos_score_list, neg_score_list, n_samp = [], [], 200
for q in samples[:n_samp]:
    pos_score_list.append(ft_model.predict([[q['query'], q['positive'][0]]])[0])
    neg_score_list.append(ft_model.predict([[q['query'], q['negative'][0]]])[0])

print('AFTER  positive avg score', sum(pos_score_list)/len(pos_score_list), 'nagative avg score', sum(neg_score_list)/len(neg_score_list), sep='\t')

AFTER  positive avg score	10.22114202260971	nagative avg score	-10.52345654964447


In [7]:
# сохраним модель и на хаб тоже:
from huggingface_hub import notebook_login
# token: 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
notebook_login()

In [51]:
# read local model
cross_ft_model = CrossEncoder("model\\crerankingeval-30e-4000-ms-marco-MiniLM-L-6-v2")
print('local', cross_ft_model.predict([("Push model","to HuggingFace")]))
# push local model to hub (tokenizer would not be uploaded to HF)
cross_ft_model.model.push_to_hub("satyroffrost/crerankingeval-30e-4000-ms-marco-MiniLM-L-6-v2")
# push tokenizer separately to hub
cross_ft_model.tokenizer.push_to_hub("satyroffrost/crerankingeval-30e-4000-ms-marco-MiniLM-L-6-v2")
# check if HF model works:
cross_ft_model = CrossEncoder("satyroffrost/crerankingeval-30e-4000-ms-marco-MiniLM-L-6-v2")
print('hf model', cross_ft_model.predict([("Push model","to HuggingFace")]))

local [10.551465]


hf model [10.551465]
